SPEC/FILESv2.md

```diff
@@ -0,0 +1,237 @@
+
+#### 1.2 Changes to returned values
+
+1. Importing a single file will now yield two entries, one for the imported file and one for the containing directory. Note this change can be considered almost backwards compatible; in the current API you'd receive an array of one value which you would access like `files[0]`. If you collect the entries in the new API you'd still access it like that.
+2. Instead of a `hash` property, entries will instead have a `cid` property. In entries yielded from core it will be a CID instance, not a string (as agreed in [ipfs/interface-js-ipfs-core#394](https://github.com/ipfs/interface-js-ipfs-core/issues/394)). In the HTTP API/CLI it will necessarily be a string, encoded in base32 by default or whatever `?cid-base`/`--cid-base` option value was requested.
+
+Example:
+
+```js
+{
+  path: '/imports/2019103114555959/myfile',
+  cid: CID,
+  size: 1234
+},
+{
+  path: '/imports/2019103114555959',
+  cid: CID,
+  size: 1234
+}
+```
+
+In js-ipfs there is a restriction when importing that disallows multiple root directories. This change would remove this restriction since there will always be a common root.
+
+#### 1.3 Remove `pin` option
+
+If users actually want to pin the data _as well_ they should use the pinning API after importing. Users may wish to do this in order to ensure the original files are retained by their node in the case where the imported files are changed.
+
+#### 1.4 Remove `wrap-with-directory` option
+
+Every import will effectively be "wrapped" in a directory so this option is no longer required.
+
+### 2. Rename `get` to `export`
+
+Renaming to `export` will more accurately describe the intention of the method.
+
+### 3. Remove `cat`
+
+For the people to whom cat means 🐈, the API method is removed and `read` used instead, which is the more obvious opposite to `write` anyway and supports IPFS paths.
+
+### 4. Hoist all methods in the `files` namespace to the root level
+
+Methods that are integral for interacting with the InterPlanetary File System will reside on the root namespace. The reasoning is that these commands are important and will be used often so need to be given priority and ease of access over other APIs that IPFS exposes. It will more effectively advertise what the IPFS core functionality is to aid onboarding and understanding of IPFS in general.
+
+For clarity, the API movement/renaming changes are as follows:
+
+| Old name | New name |
+|---|---|
+| `ipfs add` | `ipfs import` |
+| `ipfs cat` | (removed) |
+| `ipfs get` | `ipfs export` |
+| `ipfs ls` | (removed) |
+| `ipfs files cp` | `ipfs cp` |
+| `ipfs files flush` | `ipfs flush` |
+| `ipfs files ls` | `ipfs ls` |
+| `ipfs files mkdir` | `ipfs mkdir` |
+| `ipfs files mv` | `ipfs mv` |
+| `ipfs files read` | `ipfs read` |
+| `ipfs files rm` | `ipfs rm` |
+| `ipfs files stat` | `ipfs stat` |
+| `ipfs files write` | `ipfs write` |
```

Comment:
The js `files.write` implementation takes a `strategy` parameter that can be `balanced` or `trickle` (because `ipfs.add` and `ipfs.files.write` both use the same `unixfs-importer` which receives the parameter).

If you `files.write` a file, it'll get converted into a `trickle` DAG, unless you pass `strategy=balanced` in which case it'll convert it into a `balanced` DAG (rebalancing as necessary).

I think we could standardise on one and allow the user to override as they see fit.

If we could then extend the [UnixFS v1.5 metadata proposal](https://github.com/ipfs/specs/pull/223) to include the originally selected DAG strategy we'd be able to optimise for appending to trickle DAGs.